### PR TITLE
✨ Add LayerZero and Hyperlane gas estimation

### PR DESCRIPTION
**Pull request by zkSoju**
This PR adds:

- Hyperlane gas estimation with the JS SDK using FFI
- LayerZero gas estimation using the LZ endpoint

We could migrate all gas estimations and future estimations into JS scripts so it's cleaner.

To do:
- [x] Upgrade Hyperlane to V2 because SDK `DomainIdToChainName` uses V2 mapping. 
- [ ] Make output more readible.
- [x] Make gas estimates opt-in, so tests are runnable without compiling and enabling FFI

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,27 +16,46 @@ To install with [**Foundry**](https://github.com/gakonst/foundry):
...
 
-In short, it looks like this (for Hyperlane) :
+In short, it looks like this (for Hyperlane):
 
 ```js
 vm.recordLogs();
 _someCrossChainFunctionInYourContract(L2_DOMAIN, TypeCasts.addressToBytes32(address(target)));
 Vm.Log[] memory logs = vm.getRecordedLogs();
-hyperlaneHelper.help(0x35231d4c2D8B8ADcB5617A638A0c4548684c7C70, L2_FORK_ID, logs);
+hyperlaneHelper.help(L1_DOMAIN, 0x35231d4c2D8B8ADcB5617A638A0c4548684c7C70, L2_FORK_ID, logs);
```
Comment (exp-table) [inline README.md:47]: The `L1_DOMAIN` and mailbox address order is inversed. The prototype of the function is `help(address mailbox, uint32 originDomain, ...)`.

File: src/hyperlane/HyperlaneHelper.sol [hunk 2]
```diff
@@ -11,15 +11,25 @@ interface IMessageRecipient {
 }
 
 contract HyperlaneHelper is Test {
-    function help(address mailbox, uint256 forkId, Vm.Log[] calldata logs) external {
-        _help(mailbox, 0x769f711d20c679153d382254f59892613b58a97cc876b249134ac25c80f9c814, forkId, logs);
+    function help(address mailbox, uint32 originDomain, uint256 forkId, Vm.Log[] calldata logs) external {
```
Comment (exp-table) [inline src/hyperlane/HyperlaneHelper.sol:14]: Given that the `originDomain` only matters when gas is estimated, I think `originDomain` should be removed from the `help` functions and instead provide a `helpAndEstimate` function.

For the `help` functions, they would still call `_help(address mailbox, uint32 originDomain, ...)` but pass a bogus `originDomain`.

This is for both HL and LZ/SG.

`_help` could include a boolean arg to whether estimate or not.

```js
function helpWithEstimate(....) {
    bool enableEstimates = vm.envOr("ENABLE_ESTIMATES", false);
    _help(...., enableEstimates);
}

File: test/Hyperlane.t.sol [hunk 3]
```diff
@@ -42,8 +42,11 @@ contract HyperlaneHelperTest is Test {
 
     uint256 L1_FORK_ID;
     uint256 L2_FORK_ID;
+
     uint32 constant L1_DOMAIN = 1;
     uint32 constant L2_DOMAIN = 137;
+    address constant L1_hlMailbox = 0x35231d4c2D8B8ADcB5617A638A0c4548684c7C70;
```
Comment (exp-table) [inline test/Hyperlane.t.sol:48]: I would change it to `L1_HLMailbox` for readibility purposes!

File: README.md [hunk 4]
```diff
@@ -16,27 +16,46 @@ To install with [**Foundry**](https://github.com/gakonst/foundry):
...
+
+```sh
+ETH_MAINNET_RPC_URL=
+POLYGON_MAINNET_RPC_URL=
+
+# Optional
+ENABLE_ESTIMATES=
+```
+
+To enable gas estimation, from the [utils/scripts directory](./utils/scripts), run the following command:
```
Comment (exp-table) [inline README.md:29]: Should precise whether it estimates gas costs or gas consumption.

File: src/hyperlane/HyperlaneHelper.sol (hunk 2 above)
Comment (zkSoju) [reply src/hyperlane/HyperlaneHelper.sol:14]: Sounds good! I was thinking the same as well.

**Comment by zkSoju** [comment]
@exp-table Made the requested changes, but it is running into a `Stack too deep` error likely caused by adding `bool enableEstimates`. Any idea how to fix?

```
    function _help(
        address endpoint,
        address defaultLibrary,
        uint256 gasToSend,
        bytes32 eventSelector,
        uint256 forkId,
        Vm.Log[] memory logs,
        bool enableEstimates
    ) internal {
```

**Comment by exp-table** [comment]
LGTM

**Pull request merged**